libeos-updater-util/flatpak.c

```diff
@@ -1457,6 +1457,651 @@ euu_filter_for_existing_flatpak_ref_actions (GHashTable *ref_actions  /* (elemen
+                                              GCancellable         *cancellable,
+                                              GError              **error)
+{
+  g_autoptr(GPtrArray) remotes = NULL;
+  g_autoptr(GHashTable) refs_for_remotes = NULL;
+  guint i = 0;
+
+  g_return_val_if_fail (out_remotes_priority_order != NULL, FALSE);
+  g_return_val_if_fail (out_refs_for_remotes != NULL, FALSE);
+
+  remotes = flatpak_installation_list_remotes (installation, cancellable, error);
+
+  if (remotes == NULL)
+    return FALSE;
+
+  refs_for_remotes = g_hash_table_new_full (flatpak_remote_hash,
+                                            flatpak_remote_equal,
+                                            g_object_unref,
+                                            (GDestroyNotify) g_ptr_array_unref);
+
+  for (i = 0; i < remotes->len; ++i)
+    {
+      FlatpakRemote *remote = g_ptr_array_index (remotes, i);
+      g_autoptr(GPtrArray) refs_for_remote =
+        flatpak_installation_list_remote_refs_sync (installation,
+                                                    flatpak_remote_get_name (remote),
+                                                    cancellable,
+                                                    error);
+
+      if (refs_for_remote == NULL)
+        return FALSE;
+
+      g_hash_table_insert (refs_for_remotes,
+                           g_object_ref (remote),
+                           g_steal_pointer (&refs_for_remote));
+    }
+
+  /* We return both a GPtrArray of remotes and the hashtable mapping
+   * each remote to a list of refs, since we need to maintain the priority
+   * order. */
+  *out_remotes_priority_order = g_steal_pointer (&remotes);
+  *out_refs_for_remotes = g_steal_pointer (&refs_for_remotes);
+
+  return TRUE;
+}
+
+typedef struct _EuuFlatpakRelatedRefsForRemote {
+  FlatpakRemote *remote;
+  GHashTable    *related_refs;
+} EuuFlatpakRelatedRefsForRemote;
+
+static EuuFlatpakRelatedRefsForRemote *
+euu_flatpak_related_refs_for_remote_new (FlatpakRemote *remote,
+                                         GHashTable    *related_refs)
+{
+  EuuFlatpakRelatedRefsForRemote *related_refs_for_remote = g_new0 (EuuFlatpakRelatedRefsForRemote, 1);
+
+  related_refs_for_remote->remote = g_object_ref (remote);
+  related_refs_for_remote->related_refs = g_hash_table_ref (related_refs);
+
+  return related_refs_for_remote;
+}
+
+static void
+euu_flatpak_related_refs_for_remote_free (EuuFlatpakRelatedRefsForRemote *related_refs_for_remote)
+{
+  g_clear_object (&related_refs_for_remote->remote);
+  g_clear_pointer (&related_refs_for_remote->related_refs, (GDestroyNotify) g_hash_table_unref);
+
+  g_free (related_refs_for_remote);
+}
+
+/* The runtime ref is a special case that we cannot get from
+ * flatpak_installation_fetch_remote_metadata_sync. We need to manually
+ * fetch the metadata for each remote (this will usually come from the cache)
+ * and parse it for the runtime ref. */
```

Comment:
Any chance of upstreaming this to libflatpak?

Reply:
Yeah I tried to do that in https://github.com/flatpak/flatpak/pull/1578 but Alex didn't like the idea of putting it with the remote refs. He mentioned that there is a longer term goal to export such an API flatpak/flatpak#1234

Reply:
⇒ Add that as a `FIXME` comment, with links.